### DISCUSSION TITLE
Running scripts that don't exit dependent servers ?

### DISCUSSION THREAD

**Discussion by dshamim** [batect/batect#1042] [Q&A] [answered]
I came across a requirement to seed a database, 
So basically have 2 questions

I setup a seed bash script and ran it from a separate image that was dependent on the database image

They are all setup as single task - run-local-db

Question 1:
The issue I have is, as soon as the seed script finishes its run, it also closes all other dependent images
(I did get around this by pausing the script , read -p), so what's the best recommended approach for this type of setup to seed a local database ?
(assuming we want to seed it as soon as local instance of DB is started - instead of a separate task )

Question 2:
is there a way to run batect commands in detached mode  ?
e.g. when creating aws table on local dynamo - it gives an output which just pauses the script (I got around this by suppressing the output from bash)

Any recommendations on best approaches on this ?

Thanks

**Comment by charleskorn** [comment, marked as answer]
Hi @dshamim, thanks for the questions!

> Question 1:
> The issue I have is, as soon as the seed script finishes its run, it also closes all other dependent images
> (I did get around this by pausing the script , read -p), so what's the best recommended approach for this type of setup to seed a local database ?
> (assuming we want to seed it as soon as local instance of DB is started - instead of a separate task )

One approach would be to use a [setup command](https://batect.dev/docs/reference/config/containers/#setup_commands). This allows you to run a command in a container once that container has reported as healthy. So, in this case, once your database has started, you could run a command in the database container to seed the database. This does mean that you'd need to have the tools to support this available in your database container - #286 is a feature I'd like to implement one day to allow running setup commands from other containers.

> Question 2:
> is there a way to run batect commands in detached mode ?
> e.g. when creating aws table on local dynamo - it gives an output which just pauses the script (I got around this by suppressing the output from bash)

This isn't something that Batect supports - it is designed around the idea of tasks. If you'd like to start Dynamo and then wait until you kill the task, a few options come to mind:

* if your Dynamo container is the only container you're running, then that can be the task's main container and then the task will just continue running until you kill it
* if you need to run multiple containers, you'll need something that waits until you kill it, such as an application that uses Dynamo or even just [a Bash script that waits until you kill it with <kbd>Ctrl</kbd>-<kbd>C</kbd>](https://stackoverflow.com/questions/36991903/idle-bash-script-until-ctrlc-event-is-logged)

**Reply by dshamim** [reply]
Hi @charleskorn 
> 
> > Question 1:
> > The issue I have is, as soon as the seed script finishes its run, it also closes all other dependent images
> > (I did get around this by pausing the script , read -p), so what's the best recommended approach for this type of setup to seed a local database ?
> > (assuming we want to seed it as soon as local instance of DB is started - instead of a separate task )
> 
> One approach would be to use a [setup command](https://batect.dev/docs/reference/config/containers/#setup_commands). This allows you to run a command in a container once that container has reported as healthy. So, in this case, once your database has started, you could run a command in the database container to seed the database. This does mean that you'd need to have the tools to support this available in your database container - #286 is a feature I'd like to implement one day to allow running setup commands from other containers.
> 

Yea this is kind of what I'm already doing, I run the seeder, which runs in another image using command that opens a shell entry, and using volume mount that then points to a bash script to create the dynamo table and load data from a file , this has a dependency setup on the database, so that starts first

> > Question 2:
> > is there a way to run batect commands in detached mode ?
> > e.g. when creating aws table on local dynamo - it gives an output which just pauses the script (I got around this by suppressing the output from bash)
> 
> This isn't something that Batect supports - it is designed around the idea of tasks. If you'd like to start Dynamo and then wait until you kill the task, a few options come to mind:
> 
> * if your Dynamo container is the only container you're running, then that can be the task's main container and then the task will just continue running until you kill it
> * if you need to run multiple containers, you'll need something that waits until you kill it, such as an application that uses Dynamo or even just [a Bash script that waits until you kill it with Ctrl-C](https://stackoverflow.com/questions/36991903/idle-bash-script-until-ctrlc-event-is-logged)

Yep, I guess that has to be this way, I'm using an input to stop the seed image from stopping, if that closes , it takes down all dependencies with it

Thanks for your thoughts on this!
